61544 - Added read, browse and clear as communication styles to JMSSampler

bennyvw opened the pull request:
## Summary
Remarks from @pmouawad processed. New commits reflect our changes.

## Description
In JMS Point-to-Point Sampler, besides the standard options for communication style, Request Only and Request Response, we added options to Read a message, Clear the queue (purge all messages) and Browse the queue (count number of messages on the queue).
The property <stringProp> JMSSampler.connectionmode is replaced by an <intProp> JMSSampler.communicationstyle. This way, multiple languages for the options are supported.

## Motivation and Context
It does not solve a problem, but adds some functionality.

## How Has This Been Tested?
This addition has been tested using all possible options for Communication Style. All options are working correctly with IBM MQ. Also the option for reading multiple messages from a queue has been tested and works.

We did not succeed in getting ant test working without failures. Philippe, maybe you can give it a try?

This change is NOT backwards compatible in that it will not find the intProp communicationStyle in existing scripts. All scripts containing the JMS Point-to-Point sampler have to be opened and saved again to reflect this change. When the choice for Communication Style differs from 'Request Only', then the correct option has to be selected.

## Screenshots (if appropriate):
Updated screenshot in Component Reference.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.

Commit:
Added jms communication styles (read, browse and clear).

Commit:
Updated documentation with Read, Browse, Clear actions for the JMSSampler.
Renamed connectionMode to communicationStyle.

Commit:
Fixed compilation issues (removed unused packages).

Commit:
Changed formatting for the sake of checkstyle

pmouawad commented on docs/usermanual/component_reference.html:
```diff
@@ -3813,7 +3813,7 @@ <h3 id="JMS_Publisher_parms1">
   
```
HTML files are generated from XML. Could you update comonent_reference.xml instead ?

pmouawad commented on docs/usermanual/component_reference.html:
```diff
@@ -4276,10 +4289,32 @@ <h3 id="JMS_Point-to-Point_parms1">
+<dt>
+<span class="code">Browse</span>
+</dt>
+<dd> will determine the current queue depth without removing messages from the queue, returning the number of messages on the queue. 
```
Would it be possible to provide screenshots as it makes documentation much clearer

pmouawad commented on src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/JMSSampler.java:
```diff
@@ -142,48 +173,244 @@ public SampleResult sample(Entry entry) {
 
+    private void handleBrowse(SampleResult res) throws JMSException {
+        LOGGER.debug("isBrowseOnly");
+        StringBuffer sb = new StringBuffer("");
```
StringBuilder would be better here, and wherever StringBuffer is used

pmouawad commented on src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/JMSSampler.java:
```diff
@@ -142,48 +173,244 @@ public SampleResult sample(Entry entry) {
+        }
+    }
+
+    private String browseQueueForConsumption(Queue queue, String jmsSelector, SampleResult res, StringBuilder buffer,
```
Isn't name  confusing here ? AFAIU we are consuming here not browsing right ?

pmouawad commented on src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/JMSSampler.java:
```diff
@@ -142,48 +173,244 @@ public SampleResult sample(Entry entry) {
+            QueueReceiver consumer = session.createReceiver(queue, jmsSelector);
+            Message reply = consumer.receive(Long.valueOf(getTimeout()));
+            LOGGER.debug("Message: " + reply);
+            consumer.close();
```
You should use try with resource pattern to ensure close happens ? Or use try/finally and closeQuietly depending on what you want to do

pmouawad commented on src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/JMSSampler.java:
```diff
@@ -142,48 +173,244 @@ public SampleResult sample(Entry entry) {
+                res.setResponseMessage("No message received");
+            }
+        } catch (Exception ex) {
+            ex.printStackTrace();
```
This must be removed

pmouawad commented on src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/JMSSampler.java:
```diff
@@ -142,48 +173,244 @@ public SampleResult sample(Entry entry) {
+            }
+        } catch (Exception ex) {
+            ex.printStackTrace();
+            LOGGER.error(ex.getMessage());
```
Any logging requires some contextual information otherwise it's useless.

pmouawad commented on src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/JMSSampler.java:
```diff
@@ -142,48 +173,244 @@ public SampleResult sample(Entry entry) {
+            return (messageBodies + queue.getQueueName() + " has " + numMsgs + " messages");
+        } catch (Exception e) {
+            res.setResponseMessage("Error counting message on the queue");
+            e.printStackTrace();
```
Same remarks as above

pmouawad commented on src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/JMSSampler.java:
```diff
@@ -142,48 +173,244 @@ public SampleResult sample(Entry entry) {
+                deletedMsg = consumer.receiveNoWait();
+                if (deletedMsg != null) {
+                    deletedMsgCount++;
+                    deletedMsg.acknowledge();
```
Shouldn't this be parameterized ? There are 4 modes, if we ACK for AUTO_ACK that would be wrong no ?

pmouawad reviewed (changes requested):
Thanks a lot for this very interesting contribution.
Find my remarks below.
I'll be happy to merge this PR once you can take into account those remarks and possible remarks from other members.
Regards

bennyvw commented:
Philippe,

Thanks for your quick review. I am going to pay attention to your remarks asap.

Regards,

Benny.

> Op 10 november 2017 om 21:20 schreef Philippe M <notifications@github.com>:
> 
> 
>     @pmouawad requested changes on this pull request.
> 
>     Thanks a lot for this very interesting contribution.
>     Find my remarks below.
>     I'll be happy to merge this PR once you can take into account those remarks and possible remarks from other members.
>     Regards
> 
> 
>     ---------------------------------------------
> 
>     In docs/usermanual/component_reference.html https://github.com/apache/jmeter/pull/325#discussion_r150327021 :
> 
>     > @@ -3813,7 +3813,7 @@ <h3 id="JMS_Publisher_parms1">
>        
> 
>     HTML files are generated from XML. Could you update comonent_reference.xml instead ?
> 
> 
>     ---------------------------------------------
> 
>     In docs/usermanual/component_reference.html https://github.com/apache/jmeter/pull/325#discussion_r150327166 :
> 
>     > +    queue referenced by <span class="code">message.getJMSReplyTo()</span>.
>     +</dd>
>     +
>     +<dt>
>     +<span class="code">Read</span>
>     +</dt>
>     +<dd> will read a message from an outgoing queue which has no listeners attached. This can be convenient for testing purposes.
>     +     This method can be used if you need to handle queues without a binding file (in case the jmeter-jms-skip-jndi library is used),
>     +     which only works with the JMS Point-to-Point sampler.
>     +     In case binding files are used, one can also use the JMS Subscriber Sampler for reading from a queue.
>     +</dd>
>     +
>     +<dt>
>     +<span class="code">Browse</span>
>     +</dt>
>     +<dd> will determine the current queue depth without removing messages from the queue, returning the number of messages on the queue. 
> 
>     Would it be possible to provide screenshots as it makes documentation much clearer
> 
> 
>     ---------------------------------------------
> 
>     In src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/JMSSampler.java https://github.com/apache/jmeter/pull/325#discussion_r150327668 :
> 
>     >                  res.setResponseMessage(e.getLocalizedMessage());
>                  }
>              }
>              res.sampleEnd();
>              return res;
>          }
>      
>     +    private void handleBrowse(SampleResult res) throws JMSException {
>     +        LOGGER.debug("isBrowseOnly");
>     +        StringBuffer sb = new StringBuffer("");
> 
>     StringBuilder would be better here, and wherever StringBuffer is used
> 
> 
>     ---------------------------------------------
> 
>     In src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/JMSSampler.java https://github.com/apache/jmeter/pull/325#discussion_r150328109 :
> 
>     > +                Integer.parseInt(getPriority()), Long.parseLong(getExpiration()));
>     +        res.setRequestHeaders(Utils.messageProperties(msg));
>     +        if (replyMsg == null) {
>     +            res.setResponseMessage("No reply message received");
>     +        } else {
>     +            if (replyMsg instanceof TextMessage) {
>     +                res.setResponseData(((TextMessage) replyMsg).getText(), null);
>     +            } else {
>     +                res.setResponseData(replyMsg.toString(), null);
>     +            }
>     +            res.setResponseHeaders(Utils.messageProperties(replyMsg));
>     +            res.setResponseOK();
>     +        }
>     +    }
>     +
>     +    private String browseQueueForConsumption(Queue queue, String jmsSelector, SampleResult res, StringBuilder buffer,
> 
>     Isn't name confusing here ? AFAIU we are consuming here not browsing right ?
> 
> 
>     ---------------------------------------------
> 
>     In src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/JMSSampler.java https://github.com/apache/jmeter/pull/325#discussion_r150328248 :
> 
>     > +            } else {
>     +                res.setResponseData(replyMsg.toString(), null);
>     +            }
>     +            res.setResponseHeaders(Utils.messageProperties(replyMsg));
>     +            res.setResponseOK();
>     +        }
>     +    }
>     +
>     +    private String browseQueueForConsumption(Queue queue, String jmsSelector, SampleResult res, StringBuilder buffer,
>     +            StringBuilder propBuffer) {
>     +        String retVal = null;
>     +        try {
>     +            QueueReceiver consumer = session.createReceiver(queue, jmsSelector);
>     +            Message reply = consumer.receive(Long.valueOf(getTimeout()));
>     +            LOGGER.debug("Message: " + reply);
>     +            consumer.close();
> 
>     You should use try with resource pattern to ensure close happens ? Or use try/finally and closeQuietly depending on what you want to do
> 
> 
>     ---------------------------------------------
> 
>     In src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/JMSSampler.java https://github.com/apache/jmeter/pull/325#discussion_r150328288 :
> 
>     > +        try {
>     +            QueueReceiver consumer = session.createReceiver(queue, jmsSelector);
>     +            Message reply = consumer.receive(Long.valueOf(getTimeout()));
>     +            LOGGER.debug("Message: " + reply);
>     +            consumer.close();
>     +            if (reply != null) {
>     +                res.setResponseMessage("1 message(s) received successfully");
>     +                res.setResponseHeaders(reply.toString());
>     +                TextMessage msg = (TextMessage) reply;
>     +                retVal = msg.getText();
>     +                extractContent(buffer, propBuffer, msg);
>     +            } else {
>     +                res.setResponseMessage("No message received");
>     +            }
>     +        } catch (Exception ex) {
>     +            ex.printStackTrace();
> 
>     This must be removed
> 
> 
>     ---------------------------------------------
> 
>     In src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/JMSSampler.java https://github.com/apache/jmeter/pull/325#discussion_r150328360 :
> 
>     > +            QueueReceiver consumer = session.createReceiver(queue, jmsSelector);
>     +            Message reply = consumer.receive(Long.valueOf(getTimeout()));
>     +            LOGGER.debug("Message: " + reply);
>     +            consumer.close();
>     +            if (reply != null) {
>     +                res.setResponseMessage("1 message(s) received successfully");
>     +                res.setResponseHeaders(reply.toString());
>     +                TextMessage msg = (TextMessage) reply;
>     +                retVal = msg.getText();
>     +                extractContent(buffer, propBuffer, msg);
>     +            } else {
>     +                res.setResponseMessage("No message received");
>     +            }
>     +        } catch (Exception ex) {
>     +            ex.printStackTrace();
>     +            LOGGER.error(ex.getMessage());
> 
>     Any logging requires some contextual information otherwise it's useless.
> 
> 
>     ---------------------------------------------
> 
>     In src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/JMSSampler.java https://github.com/apache/jmeter/pull/325#discussion_r150328441 :
> 
>     > +                if (corrID == null) {
>     +                    corrID = message.getJMSMessageID();
>     +                    messageBodies = messageBodies + numMsgs + " - MessageID: " + corrID + ": " + message.getText()
>     +                            + "\n";
>     +                } else {
>     +                    messageBodies = messageBodies + numMsgs + " - CorrelationID: " + corrID + ": " + message.getText()
>     +                            + "\n";
>     +                }
>     +                numMsgs++;
>     +            }
>     +            res.setResponseMessage(numMsgs + " messages available on the queue");
>     +            res.setResponseHeaders(qBrowser.toString());
>     +            return (messageBodies + queue.getQueueName() + " has " + numMsgs + " messages");
>     +        } catch (Exception e) {
>     +            res.setResponseMessage("Error counting message on the queue");
>     +            e.printStackTrace();
> 
>     Same remarks as above
> 
> 
>     ---------------------------------------------
> 
>     In src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/JMSSampler.java https://github.com/apache/jmeter/pull/325#discussion_r150328999 :
> 
>     > +            LOGGER.error(e.getMessage());
>     +            return "";
>     +        }
>     +    }
>     +
>     +    private String clearQueue(Queue queue, SampleResult res) {
>     +        String retVal = null;
>     +        try {
>     +            QueueReceiver consumer = session.createReceiver(queue);
>     +            Message deletedMsg = null;
>     +            long deletedMsgCount = 0;
>     +            do {
>     +                deletedMsg = consumer.receiveNoWait();
>     +                if (deletedMsg != null) {
>     +                    deletedMsgCount++;
>     +                    deletedMsg.acknowledge();
> 
>     Shouldn't this be parameterized ? There are 4 modes, if we ACK for AUTO_ACK that would be wrong no ?
> 
>     —
>     You are receiving this because you authored the thread.
>     Reply to this email directly, view it on GitHub https://github.com/apache/jmeter/pull/325#pullrequestreview-75868433 , or mute the thread https://github.com/notifications/unsubscribe-auth/Aem4Xh4JYZs7qs0fAg01NRZYm8aTa2g-ks5s1LAlgaJpZM4QZnSl .
> 
>      
> 



Met vriendelijke groet,

Benny van Wijngaarden.
Tel. 0629556587http://www.smaragd-it.nl/


benny@smaragd-it.nl mailto:benny@smaragd-it.nl

Commit:
component_reference.html changes reverted. xml edited to contain the changes. Default value for dropdown list changed into Request Only.

Commit:
Refactored JMSSampler: some renaming and handleRead method split into smaller methods.

Commit:
Screenshot JMS Point-to-Point sampler updated with field Number of samples to aggregate

Commit:
Merge branch '61544' of https://github.com/bennyvw/jmeter into 61544

# Conflicts:
#	src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/JMSSampler.java

Commit:
Processed review comments from Philippe Mouawad.

Commit:
Merge branch '61544' of https://github.com/bennyvw/jmeter into 61544

Commit:
Refactored JMSSampler: split large methods into smaller methods.

bennyvw commented:
Processed your remarks, @pmouawad. Could you please review it again? And you might give us some hint why tests are failing in Travis CI. It complains about JMS_TESTS.JMX, but I did not find that file..

pmouawad commented:
Hello @bennyvw ,
Thanks for updating PR.
It seems the changes have introduced an issue in JMS_TESTS.jmx which is located in source files on github/svn (only, not in binary) in bin/testfiles:
https://github.com/apache/jmeter/blob/trunk/bin/testfiles/JMS_TESTS.jmx

To debug the issue, put activemq-all.jar in lib folder and run the test.

Thanks

bennyvw committed:
Added JMS Tests (copied from trunk), and receiveNoWait() changed to receive(Long) in clear function

bennyvw committed:
Merge branch 'trunk' into 61544

bennyvw commented:
Philippe,

In my branch 61544, I did not have the JMS_TESTS files (jmx, xml, csv). I don't know why, do you?

I added them manually from trunk, even after merging trunk into 61544 I did not have them.

The issues that were introduced in those tests have been solved and I have added tests for Read, Browse and Clear queue. Because Clear queue did not work on the ActiveMQ queues, I changed the code slightly.

Just solved the merge conflicts in JMS_TESTS.JMX.

Could you please give it another review?

Regards,

Benny.

> Op 14 november 2017 om 18:24 schreef Philippe M <notifications@github.com>:
> 
> 
>     Hello @bennyvw https://github.com/bennyvw ,
>     Thanks for updating PR.
>     It seems the changes have introduced an issue in JMS_TESTS.jmx which is located in source files on github/svn (only, not in binary) in bin/testfiles:
>     https://github.com/apache/jmeter/blob/trunk/bin/testfiles/JMS_TESTS.jmx
> 
>     To debug the issue, put activemq-all.jar in lib folder and run the test.
> 
>     Thanks
> 
>     —
>     You are receiving this because you were mentioned.
>     Reply to this email directly, view it on GitHub https://github.com/apache/jmeter/pull/325#issuecomment-344333265 , or mute the thread https://github.com/notifications/unsubscribe-auth/Aem4XtE1GQcFri2ghrxZpyxu9skZZjiFks5s2czhgaJpZM4QZnSl .
> 
>      
> 



Met vriendelijke groet,

Benny van Wijngaarden.
Tel. 0629556587http://www.smaragd-it.nl/


benny@smaragd-it.nl mailto:benny@smaragd-it.nl

Commit:
Added expected lines to CSV for new JMS tests

Commit:
Merge branch '61544' of https://github.com/bennyvw/jmeter into 61544

Commit:
Added expected lines to XML results for new JMS tests

Commit:
Forgot newline at end of file. Now added.

Commit:
Forgot newline at end of file. Now added.
